### DISCUSSION TITLE
How to handle `uv_write_t` request allocation correctly?

### DISCUSSION THREAD

**Discussion by Oldes** [libuv/libuv#4501] [Q&A]
Hi...
I'm new to `libuv` and I want to use a simple function to write into the stderr pipe and have function like this:
```C
void on_err_write(uv_write_t *req, int status) {
    free(req);
}
void write_error(const char*str){
    uv_buf_t buffer = uv_buf_init((char*)str, strlen(str));
    uv_write_t *req = (uv_write_t*)malloc(sizeof(uv_write_t));
    uv_write(req, (uv_stream_t*)&stderr_pipe, &buffer, 1, on_write_error);
}
```
It works but I wonder if there always must be the request allocated? If I want to avoid these allocations on each call, do I have to handle own pool of requests or is there any built in solution?

Is it safe to use a global request? Like:
```C
uv_write_t error_req;
void write_error(const char*str){
    uv_buf_t buffer = uv_buf_init((char*)str, strlen(str));
    uv_write(&error_req, (uv_stream_t*)&stderr_pipe, &buffer, 1, NULL);
}
```
Which also opens another question, if there is no callback used, is the call still evaluated asynchronously?

**Comment by Oldes** [comment]
Ok.. it looks that I can answer on the second part myself.. it is not save to use a global request :-)

**Comment by saghul** [comment]
You need a fresh request each time, but you could keep a free-list and reuse them, as long as you make sure a given request is only reused after its callback was called.

**Comment by Oldes** [comment]
Thanks @saghul, what about the `buffer` in my code above? Is it safe when it is a function local or should it be also allocated?

**Reply by saghul** [reply]
The buffer is OK, the str that you initialized it needs to live at least until the write callback is called.
